File: 124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int res;
    int dfs(TreeNode *root)
    {
        if (root == NULL)
        {
            return 0;
        }
        int leftMax = max(dfs(root->left), 0);
        int rightMax = max(dfs(root->right), 0);
        res = max(res, root->val + leftMax + rightMax);
        return root->val + max(leftMax, rightMax);
    }
    int maxPathSum(TreeNode *root)
    {

        res = root->val;

        dfs(root);
        return res;
    }
};

File: 124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = root.val
        def traverse(root):
            nonlocal res
            if root == None:
                return 0
            
            leftMax = max(traverse(root.left), 0)
            rightMax = max(traverse(root.right), 0)

            res = max(rightMax + leftMax + root.val, res)

            return root.val + max(rightMax, leftMax)
        
        traverse(root)
        return res
